### ISSUE TITLE
`needless_pass_by_ref_mut` false positive because of type change

### ISSUE THREAD

**Issue by smoelius** [rust-lang/rust-clippy#11182]
### Summary

A function may require an `&mut` argument so that it has a particular type.

### Lint Name

needless_pass_by_ref_mut

### Reproducer

I tried this code (note that `foo` needs to have the `&mut` argument to be passed to `bar`):

```rust
fn foo(y: &mut i32) -> i32 {
    12 + *y
}

fn bar(_: impl Fn(&mut i32) -> i32) {}

fn main() {
    bar(foo);
}
```

I saw this happen:

```
warning: this argument is a mutable reference, but not used mutably
 --> src/main.rs:1:11
  |
1 | fn foo(y: &mut i32) -> i32 {
  |           ^^^^^^^^ help: consider changing to: `&i32`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_pass_by_ref_mut
  = note: `#[warn(clippy::needless_pass_by_ref_mut)]` on by default
```

I expected to see this happen: no warning.

### Version

```text
rustc 1.73.0-nightly (da6b55cc5 2023-07-17)
binary: rustc
commit-hash: da6b55cc5eaf76ed6acb7dc2f7d611e32af7c9a7
commit-date: 2023-07-17
host: x86_64-unknown-linux-gnu
release: 1.73.0-nightly
LLVM version: 16.0.5
```

**Comment by Centri3** [comment]
cc @GuillaumeGomez

This would require some work, but should be possible to fix. My lint `single_call_fn` does something similar, use a visitor to see where the function's accessed and suppress the warning if it's passed to a function that takes `&mut`. You'd likely want to store the function `LocalDefId`s in the lint's struct then have a `check_crate_post`
